flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java

```diff
@@ -21,519 +21,897 @@
+
+		if (null != slot) {
+			// we assume the given allocation id to be the ground truth (coming from the TM)
+			slot.setAllocationId(allocationId);
+
+			if (null != allocationId) {
+				if (slot.hasPendingSlotRequest()){
+					// we have a pending slot request --> check whether we have to reject it
+					PendingSlotRequest pendingSlotRequest = slot.getAssignedSlotRequest();
+
+					if (Objects.equals(pendingSlotRequest.getAllocationId(), allocationId)) {
+						// we can cancel the slot request because it has been fulfilled
+						cancelPendingSlotRequest(pendingSlotRequest);
+
+						// remove the pending slot request, since it has been completed
+						pendingSlotRequests.remove(pendingSlotRequest.getAllocationId());
+					} else {
+						// this will try to find a new slot for the request
+						rejectPendingSlotRequest(
+							pendingSlotRequest,
+							new Exception("Task manager reported slot " + slotId + " being already allocated."));
+					}
+
+					slot.setAssignedSlotRequest(null);
+				}
+
+				fulfilledSlotRequests.put(allocationId, slotId);
+
+				TaskManagerRegistration taskManagerRegistration = taskManagerRegistrations.get(slot.getInstanceId());
+
+				if (null != taskManagerRegistration) {
+					// disable any registered time out for the task manager
+					taskManagerRegistration.cancelTimeout();
 				}
 			}
+		} else {
+			LOG.debug("Trying to update unknown slot with slot id {}.", slotId);
 		}
 	}
 
-	// ------------------------------------------------------------------------
-	//  internal behaviors
-	// ------------------------------------------------------------------------
-
 	/**
-	 * When we have a free slot, try to fulfill the pending request first. If any request can be fulfilled,
-	 * record this allocation in bookkeeping and send slot request to TaskManager, else we just add this slot
-	 * to the free pool.
+	 * Tries to allocate a slot for the given slot request. If there is no slot available, the
+	 * resource manager is informed to allocate more resources and a timeout for the request is
+	 * registered.
 	 *
-	 * @param freeSlot The free slot
+	 * @param pendingSlotRequest to allocate a slot for
+	 * @throws ResourceManagerException if the resource manager cannot allocate more resource
 	 */
-	private void handleFreeSlot(final ResourceSlot freeSlot) {
-		SlotRequest chosenRequest = chooseRequestToFulfill(freeSlot, pendingSlotRequests);
+	private void internalRequestSlot(PendingSlotRequest pendingSlotRequest) throws ResourceManagerException {
+		TaskManagerSlot taskManagerSlot = findMatchingSlot(pendingSlotRequest.getResourceProfile());
 
-		if (chosenRequest != null) {
-			final AllocationID allocationId = chosenRequest.getAllocationId();
-			final SlotRequest slotRequest = pendingSlotRequests.remove(allocationId);
+		if (taskManagerSlot != null) {
+			allocateSlot(taskManagerSlot, pendingSlotRequest);
+		} else {
+			final UUID timeoutIdentifier = UUID.randomUUID();
```

Comment:
True, indeed. I think we don't need these identifier anymore if we have a single timeout task which periodically checks all task managers and pending slot requests.